library/std/src/sys_common/once/futex.rs

```diff
@@ -0,0 +1,138 @@
+use crate::cell::Cell;
+use crate::sync as public;
+use crate::sync::atomic::{
+    AtomicU32,
+    Ordering::{Acquire, Relaxed, Release},
+};
+use crate::sys::futex::{futex_wait, futex_wake_all};
+
+// On some platforms, the OS is very nice and handles the waiter queue for us.
+// This means we only need one atomic value with 5 states:
+
+/// No initialization has run yet, and no thread is currently using the Once.
+const INCOMPLETE: u32 = 0;
+/// Some thread has previously attempted to initialize the Once, but it panicked,
+/// so the Once is now poisoned. There are no other threads currently accessing
+/// this Once.
+const POISONED: u32 = 1;
+/// Some thread is currently attempting to run initialization. It may succeed,
+/// so all future threads need to wait for it to finish.
+const RUNNING: u32 = 2;
+/// Some thread is currently attempting to run initialization and there are threads
+/// waiting for it to finish.
+const QUEUED: u32 = 3;
+/// Initialization has completed and all future calls should finish immediately.
+const COMPLETE: u32 = 4;
+
+// Threads wait by setting the state to QUEUED and calling `futex_wait` on the state
+// variable. When the running thread finishes, it will wake all waiting threads using
+// `futex_wake_all`.
+
+pub struct OnceState {
+    poisoned: bool,
+    set_state_to: Cell<u32>,
+}
+
+impl OnceState {
+    #[inline]
+    pub fn is_poisoned(&self) -> bool {
+        self.poisoned
+    }
+
+    #[inline]
+    pub fn poison(&self) {
+        self.set_state_to.set(POISONED);
+    }
+}
+
+struct WaiterQueue<'a> {
+    state: &'a AtomicU32,
+    set_state_on_drop_to: u32,
+}
+
+impl<'a> Drop for WaiterQueue<'a> {
+    fn drop(&mut self) {
+        // Use release ordering to propagate changes to all threads checking
+        // up on the Once. `futex_wake_all` does its own synchronization, hence
+        // we do not need `AcqRel`.
+        if self.state.swap(self.set_state_on_drop_to, Release) == QUEUED {
+            futex_wake_all(&self.state);
+        }
+    }
+}
+
+pub struct Once {
+    state: AtomicU32,
+}
+
+impl Once {
+    #[inline]
+    pub const fn new() -> Once {
+        Once { state: AtomicU32::new(INCOMPLETE) }
+    }
+
+    #[inline]
+    pub fn is_completed(&self) -> bool {
+        // Use acquire ordering to make all initialization changes visible to the
+        // current thread.
+        self.state.load(Acquire) == COMPLETE
+    }
+
+    // This uses FnMut to match the API of the generic implementation. As this
+    // implementation is quite light-weight, it is generic over the closure and
+    // so avoids the cost of dynamic dispatch.
+    //
+    // Also, this is tagged with `#[cold]` as it should indeed be cold and it
```

Comment:
Cold has a lot of knock-on effects that cause issues even in cases that are indeed cold/off the fast path.

Specifically, branches dominated by calls to `#[cold]` functions end up considered cold themselves, so if the common. This means that cases like Vec where the compiler can tell that the `#[cold]` growth function is always called in usage like `let mut v = Vec::new(); v.push(...);`, the compiler will sometimes incorrectly consider that caller cold, even though it's a case "this is statistically unlikely, but always happens on the first call". The main outcome of this ends up being failure to optimize Vec-using code as well as clang does with libc++'s std::vector (sadly, there's no easy fix, as removing the `#[cold]` seems to cause the compiler to predict some branches incorrectly, and we don't have working likelihood intrinsics anymore).

So, `Once` is the same kind of case as Vec, since first time through this is always called. That said, `#[cold]` here isn't be problematic, as the issue could only happen if the compiler knew what value will be loaded from `self.state` in the caller. This is probably impossible in any code that actually needed `Once`, and if it's not impossible, then that code has some performance issues anyway, due to the unnecessary use of `Once`.

I think this comment is kind of misleading as a result though. Might be better off removed, since it's pretty clear why it's `#[cold]` IMO.